File: week10/jisoo/프로그래머스 150369. 택배 배달과 수거하기.java
// 일렬로 나열된 n개 집에 택배 배달
// i번째 집은 물류창고에서 거리 i만큼 떨어져 있음, i와 j번째 집은 j-i만큼 ㅇㅇ
// 트럭에 상자를 최대 cap개 실을 수 있음
// 상자들 실어서 물류창고에서 출발해 각 집에 배달하면서, 빈 상자 수거해 물류창고에서 내림
// 트럭 하나로 모든 배달과 수거 마치고 물류창고까지 돌아올 수 있는 최소 이동 거리는?
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;
        int del = 0; int pick = 0;

        // 마지막 집부터 시작
        for (int i=n-1; i>=0; i--) {
            // 현재 집에 몇 번 방문해야 하는지 측정
            if (deliveries[i]>0 || pickups[i]>0) {
                int cnt = 0;

                while (del<deliveries[i] || pick<pickups[i]) {
                    cnt+=1;
                    del += cap;
                    pick += cap;
                }

                // 현재 집에 배달, 수거하고
                // 추가로 배달, 수거 가능한 상자 개수 구하기
                del -= deliveries[i];
                pick -= pickups[i];
                answer += (i+1)*cnt*2;
            }
        }
        return answer;
    }
}

File: week10/chanki/프로그래머스 150369. 택배 배달과 수거하기.kt
class P150369Kt {
    fun solution(cap: Int, n: Int, deliveries: IntArray, pickups: IntArray): Long {
        // 가장 멀리 떨어진 배송, 픽업 위치
        var lastDeliveredPlace = 0;
        var lastPickupPlace = 0;

        for (i in n - 1 downTo 0) {
            if (lastDeliveredPlace == 0 && deliveries[i] != 0) {
                lastDeliveredPlace = i
            }

            if (lastPickupPlace == 0 && pickups[i] != 0) {
                lastPickupPlace = i
            }

            if (lastDeliveredPlace != 0 && lastPickupPlace != 0) {
                break
            }
        }

        // 총 이동 거리
        var distance = 0L
        // 모든 배송과 픽업이 처리될 때까지
        while (deliveries[lastDeliveredPlace] > 0 || pickups[lastPickupPlace] > 0) {
            // 거리는 배송지와 픽업지 중 더 먼 곳을 기준으로 왕복하는 것이 됨
            distance += (maxOf(lastDeliveredPlace, lastPickupPlace) + 1) * 2

            // cap을 모두 꽉꽉 채워, 가장 먼 곳부터 배송을 처리
            var sumOfDeliveryPack = 0
            for (i in lastDeliveredPlace downTo 0) {
                lastDeliveredPlace = i
                if (sumOfDeliveryPack + deliveries[i] > cap) {
                    deliveries[i] -= cap - sumOfDeliveryPack
                    break
                }

                sumOfDeliveryPack += deliveries[i]
                deliveries[i] = 0
            }

            // cap이 텅 빈 상태로, 가장 먼 곳부터 픽업을 처리
            var sumOfPickupPack = 0
            for (i in lastPickupPlace downTo 0) {
                lastPickupPlace = i
                if (sumOfPickupPack + pickups[i] > cap) {
                    pickups[i] -= cap - sumOfPickupPack
                    break
                }

                sumOfPickupPack += pickups[i]
                pickups[i] = 0
            }
        }

        return distance
    }
}